### PR TITLE
first pass at redesigning submit

### PR DESCRIPTION
**Pull request by DivadNojnarg**
- Improved `submit` feature for blocks. Now submit isn't added as a class but as an optional block attribute. When you design a block, you can pass the `submit` parameter like so:

```r
new_super_block <- function(submit = 0, ...) {
  fields <- list()
  new_block(
    fields = fields,
    expr = quote(print("test")),
    submit = submit,
    ...,
    class = "my_block"
  )
}
```

When `submit = 0`, it will add a submit button in the block toolbar but computations are blocked (clicking on it is required). Internally, each time the `input$submit` is clicked, the submit attribute is incremented by 1 to stay in sync with the UI button. This is useful when the stack is serialised, since this state is kept so that computations can be automatically re-triggered on restore, if `submit > 0`.

```r
serve_stack(new_stack(new_dataset_block(), new_filter_block(columns = "Time", submit = 0)))

# Simulate what happens when restoring a serialised stack
# with submit > 0
serve_stack(new_stack(new_dataset_block(), new_filter_block(columns = "Time", submit = 1)))
```

### CODE REVIEW

**Comment by DivadNojnarg** [comment]
Made some tweaks to the button styling. I made it always visible, at the center of the block title, as default.

https://github.com/user-attachments/assets/345db89e-f3ee-47c7-9e19-e60805100d9e

**Comment by DivadNojnarg** [comment]
Button was improved to use the bslib task[^1] button that can't be clicked as long as the task is running.

https://github.com/user-attachments/assets/05db09de-dee8-4e6f-849b-07a76f1e0e83

[^1]: In the future, this can be coupled to `ExtendedTask` if we want to handle async tasks for some blocks and stacks and improve performances).

File: R/utils.R [hunk 1]
```diff
@@ -281,7 +280,7 @@ send_error_to_ui <- function(blk, is_valid, session) {
   )
 
   # Send validation message
-  if (!is_valid$block) {
+  if (!is_valid$block && length(is_valid$message)) {
```
Comment (DivadNojnarg) [inline R/utils.R:283]: Fix #415

File: R/ui.R [hunk 2]
```diff
@@ -331,7 +346,6 @@ add_block_ui.default <- function(x, id, ...) {
         tags$head(
           tags$script(HTML("
             function colorText(data) {
-              console.log(data);
```
Comment (DivadNojnarg) [inline R/ui.R:334]: Fix #417

**Comment by DivadNojnarg** [comment]
Tweaks to make button more visible. cc @JohnCoene 

Screenshot 2024-09-16 at 12 28 46

**Comment by JohnCoene** [comment]
We'll have to set the version of shiny in `DESCRIPTION` too I believe (since it relies on the latest version)

**Comment by DivadNojnarg** [comment]
@nbenn Feel free to review.

**Review by nbenn — COMMENTED** [review]
Thanks for looking into this. Some (minor) comments:

- You brought this up yourself in a private msg: should `submit` also be an argument of the "base" ctor? I think *yes* it should. It is a *special* block attribute in the sense that some core behavior depends on these values and adding this to the `new_block()` signature communicates this to a user of our API.
- As a user, I would want to be able to pass a boolean value as `submit`. If I see `submit = 0`, I don't know whether this means `TRUE` or `FALSE` and even if I know `submit = 0` <=> "submit = TRUE", I could then mistakenly assume that `submit = 1` <=> "submit = FALSE". We could allow a flag to be passed as `submit` and then in the base ctor convert `TRUE` -> `0` and `FASLE` -> `-1` or so.
- Personally I'd prefer to have an explicit value as `submit` attribute (e.g. "-1" or so) instead of it not being available corresponding to `submit = FALSE`. This could be resolved just by adding `submit = FALSE` to `new_block()`.
- This is inconsequential, so feel free to ignore: we're only really interested in 3 possible values ("no submit", "submit, but not yet triggered", "already triggered submit"). We could map this to values `c(-1, 0, 1)` or even more compactly `c(FALSE, NA, TRUE)`. Tbh, I'm not super happy w/ my suggestions here. I still feel that differentiating between a value of 543902 and 543903 is pointless. But it also does not hurt.

**Comment by DivadNojnarg** [comment]
@nbenn Included your suggestions and added some unit tests. I am stuck by a shinytest2 issue and Positron. I'd suggest to merge without waiting for this to be solved. I'll open another PR later to add more tests.

**Review by nbenn — APPROVED** [review]
Lgtm.

**Pull request merged**